<a href="https://colab.research.google.com/github/pnp642001/Income_Classification/blob/main/US_INCOME_PREDICTION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**PARTH PATEL**

##**TO CLASSIFY WHETHER THE INCOME IS LESS THAN $50K OR MORE**

In [ ]:
#First, we will convert the .csv file into data_frame using pandas

import pandas as pd

df=pd.read_csv('adult.csv')
print(df.shape)

(32560, 15)


In [ ]:
df.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [ ]:
#WE WILL ADD THE COLUMN NAMES

df.columns=['Age','Workclass','Fnlwgt','Education','education_num','marital_status','occupation','relationship','race','sex','capital_gain','capital_loss','hours_per_week','native_country','income']


In [ ]:
df.head()

Age          Workclass  Fnlwgt   Education  education_num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   
3   28            Private  338409   Bachelors             13   
4   37            Private  284582     Masters             14   

        marital_status          occupation    relationship    race      sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40            Cuba   <=50K  
4             0             0              40   United-States   <=50K

In [ ]:
#WE WILL CHECK FOR NULL VALUES
df.isnull().sum()

Age               0
Workclass         0
Fnlwgt            0
Education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [ ]:
df.dtypes

Age                int64
Workclass         object
Fnlwgt             int64
Education         object
education_num      int64
marital_status    object
occupation        object
relationship      object
race              object
sex               object
capital_gain       int64
capital_loss       int64
hours_per_week     int64
native_country    object
income            object
dtype: object

In [ ]:
#AS CERTAIN ATTRIBUTES ARE NOT NUMERIC. WE WILL NEED TO CONVERT THEM INTO NUMERICAL FORM

In [ ]:
#COLUMN NAME : Workclass, marital_status occupation relationship  race  sex  native_country income ARE OF OBJECT TYPE
# WE WILL USE LABEL ENCODER



In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
cat_columns_1= [ i for i in df.columns if df[i].dtypes=='object']
print(cat_columns_1)

['Workclass', 'Education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country', 'income']


In [ ]:
lb=LabelEncoder()

for i in cat_columns_1:
  df[i]=lb.fit_transform(df[i])

In [ ]:
#NEW DATA TYPES OF COLUMNS

df.dtypes

Age               int64
Workclass         int64
Fnlwgt            int64
Education         int64
education_num     int64
marital_status    int64
occupation        int64
relationship      int64
race              int64
sex               int64
capital_gain      int64
capital_loss      int64
hours_per_week    int64
native_country    int64
income            int64
dtype: object

In [ ]:
df.head()

Age  Workclass  Fnlwgt  Education  education_num  marital_status  \
0   50          6   83311          9             13               2   
1   38          4  215646         11              9               0   
2   53          4  234721          1              7               2   
3   28          4  338409          9             13               2   
4   37          4  284582         12             14               2   

   occupation  relationship  race  sex  capital_gain  capital_loss  \
0           4             0     4    1             0             0   
1           6             1     4    1             0             0   
2           6             0     2    1             0             0   
3          10             5     2    0             0             0   
4           4             5     4    0             0             0   

   hours_per_week  native_country  income  
0              13              39       0  
1              40              39       0  
2              40              39       0  
3              40               5       0  
4              40              39       0

In [ ]:
df['Workclass'].value_counts()

4    22696
6     2541
2     2093
0     1836
7     1297
5     1116
1      960
8       14
3        7
Name: Workclass, dtype: int64

In [ ]:
df['Education'].value_counts()

11    10501
15     7291
9      5354
12     1723
8      1382
1      1175
7      1067
0       933
5       646
14      576
6       514
2       433
10      413
4       333
3       168
13       51
Name: Education, dtype: int64

In [ ]:


df['marital_status'].value_counts()

2    14976
4    10682
0     4443
5     1025
6      993
3      418
1       23
Name: marital_status, dtype: int64

In [ ]:

df['occupation'].value_counts()

10    4140
3     4099
4     4066
1     3769
12    3650
8     3295
7     2002
0     1843
14    1597
6     1370
5      994
13     928
11     649
9      149
2        9
Name: occupation, dtype: int64

In [ ]:
df['relationship'].value_counts()

0    13193
1     8304
3     5068
4     3446
5     1568
2      981
Name: relationship, dtype: int64

In [ ]:
df['race'].value_counts()

4    27815
2     3124
1     1039
0      311
3      271
Name: race, dtype: int64

In [ ]:
df['sex'].value_counts()

1    21789
0    10771
Name: sex, dtype: int64

In [ ]:
 df['native_country'].value_counts()

39    29169
26      643
0       583
30      198
11      137
2       121
33      114
8       106
19      100
5        95
9        90
23       81
35       80
3        75
22       73
6        70
40       67
13       64
24       62
31       60
4        59
36       51
14       44
20       43
32       37
27       34
29       31
10       29
12       29
7        28
21       24
17       20
1        19
38       19
25       18
37       18
41       16
28       14
16       13
18       13
34       12
15        1
Name: native_country, dtype: int64

In [ ]:
df['income'].value_counts()

0    24719
1     7841
Name: income, dtype: int64

#**BUILDING MODEL**

In [ ]:
#Creating a custom function for reducing the redundancy for the confusion matrix and classification report
def generate_cm_accuracy(y_test,y_pred):
  from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
  cm=confusion_matrix(y_test,y_pred)
  print(cm)
  print(classification_report(y_test,y_pred))
  print("ACCURACY OF MODEL : ",accuracy_score(y_test,y_pred))
  print("----------------------------------------------------")
#WE KNOW : ACCURACY = 1-MISCLASSIFICATION
# THUS, MISCLASSIFICATION = 1-ACCURACY
# PERCENTAGE OF MISCLASSIFICATION WILL BE 100*(1-ACCURACY)
  print("PERCENTAGE OF MISCLASSIFICATION : ",100*(1-accuracy_score(y_test,y_pred)))

In [ ]:
#FUNCTION FOR TRAINING AND TESTING SCORE

def train_test_score(model):
  print("Training Score: ",model.score(x_train,y_train))
  print("Testing Score: ",model.score(x_test,y_test))

In [ ]:
#Now, we will split the given data into train and test data.
#Test size will be 20%

#Divding the data into dependent and independent categories.

x=df.drop('income',axis=1)
y=df['income']

print(type(x),type(y))
print(x.shape,y.shape)

<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.series.Series'>
(32560, 14) (32560,)


In [ ]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)

In [ ]:
print(x_train.shape,y_train.shape)

(26048, 14) (26048,)


In [ ]:
#IMPORTING THE IMPORTANT LIBRARIES FOR THE MODELS:

# WE WILL USE THE FOLLOWING ALGORITHMS:
# a. Decision Tree Classifier
# b. Random Forest Classifier
# c. KNN Classifier
# d. Logistic Regression
# e. SVM Classifier

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from  sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

#**DECISION TREE CLASSIFIER**


In [ ]:
model_1=DecisionTreeClassifier(criterion='gini',max_depth=9)
model_1.fit(x_train,y_train)



DecisionTreeClassifier(max_depth=9)

In [ ]:
train_test_score(model_1)

Training Score:  0.8638667076167076
Testing Score:  0.8521191646191646


In [ ]:
ypred_model_1=model_1.predict(x_test)
print(ypred_model_1)

[0 0 0 ... 0 0 0]


In [ ]:
generate_cm_accuracy(y_test,ypred_model_1)

[[4641  328]
 [ 635  908]]
              precision    recall  f1-score   support

           0       0.88      0.93      0.91      4969
           1       0.73      0.59      0.65      1543

    accuracy                           0.85      6512
   macro avg       0.81      0.76      0.78      6512
weighted avg       0.85      0.85      0.85      6512

ACCURACY OF MODEL :  0.8521191646191646
----------------------------------------------------
PERCENTAGE OF MISCLASSIFICATION :  14.78808353808354


#**RANDOM FOREST CLASSIFIER**

In [ ]:
model_2=RandomForestClassifier(n_estimators=100,criterion='gini',max_depth=6)
model_2.fit(x_train,y_train)

RandomForestClassifier(max_depth=6)

In [ ]:
train_test_score(model_2)

Training Score:  0.8551904176904177
Testing Score:  0.8527334152334153


In [ ]:
ypred_model_2=model_2.predict(x_test)
print(ypred_model_2)

[0 0 0 ... 0 0 0]


In [ ]:
print(generate_cm_accuracy(y_test,ypred_model_2))

[[4771  198]
 [ 761  782]]
              precision    recall  f1-score   support

           0       0.86      0.96      0.91      4969
           1       0.80      0.51      0.62      1543

    accuracy                           0.85      6512
   macro avg       0.83      0.73      0.76      6512
weighted avg       0.85      0.85      0.84      6512

ACCURACY OF MODEL :  0.8527334152334153
----------------------------------------------------
PERCENTAGE OF MISCLASSIFICATION :  14.726658476658471
None


#**KNN**

In [ ]:
#for N in Knn, N=sqrt(training_size)
# tranining size is 26048
# N=161
model_3=KNeighborsClassifier(n_neighbors=161)
model_3.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=161)

In [ ]:
train_test_score(model_3)

Training Score:  0.7793688574938575
Testing Score:  0.7811732186732187


In [ ]:
ypred_model_3=model_3.predict(x_test)
print(ypred_model_3)

[0 0 0 ... 0 0 0]


In [ ]:
print(generate_cm_accuracy(y_test,ypred_model_3))

[[4967    2]
 [1423  120]]
              precision    recall  f1-score   support

           0       0.78      1.00      0.87      4969
           1       0.98      0.08      0.14      1543

    accuracy                           0.78      6512
   macro avg       0.88      0.54      0.51      6512
weighted avg       0.83      0.78      0.70      6512

ACCURACY OF MODEL :  0.7811732186732187
----------------------------------------------------
PERCENTAGE OF MISCLASSIFICATION :  21.882678132678134
None


#**LOGISTIC REGRESSION**

In [ ]:
model_4=RandomForestClassifier()
model_4.fit(x_train,y_train)

RandomForestClassifier()

In [ ]:
train_test_score(model_4)

Training Score:  1.0
Testing Score:  0.8585687960687961


In [ ]:
ypred_model_4=model_4.predict(x_test)
print(ypred_model_4)

[0 0 0 ... 0 0 0]


In [ ]:
print(generate_cm_accuracy(y_test,ypred_model_4))

[[4639  330]
 [ 591  952]]
              precision    recall  f1-score   support

           0       0.89      0.93      0.91      4969
           1       0.74      0.62      0.67      1543

    accuracy                           0.86      6512
   macro avg       0.81      0.78      0.79      6512
weighted avg       0.85      0.86      0.85      6512

ACCURACY OF MODEL :  0.8585687960687961
----------------------------------------------------
PERCENTAGE OF MISCLASSIFICATION :  14.14312039312039
None


#**SVM**

In [ ]:
model_5=SVC(kernel='linear',probability=False)
model_5.fit(x_train,y_train)

SVC(kernel='linear')

In [ ]:
train_test_score(model_5)

Training Score:  0.7818740399385561
Testing Score:  0.7974552634053088


In [ ]:
ypred_model_5=model_5.predict(x_test)
print(ypred_model_5)

[0 0 0 ... 0 0 0]


In [ ]:
print(generate_cm_accuracy(y_test,ypred_model_5))

[[23649   584]
 [ 5879  1797]]
              precision    recall  f1-score   support

           0       0.80      0.98      0.88     24233
           1       0.75      0.23      0.36      7676

    accuracy                           0.80     31909
   macro avg       0.78      0.61      0.62     31909
weighted avg       0.79      0.80      0.75     31909

ACCURACY OF MODEL :  0.7974552634053088
----------------------------------------------------
PERCENTAGE OF MISCLASSIFICATION :  20.25447365946912
None


#**MOST ACCURATE MODEL FOR THE GIVEN DATA-SET:**



* ACCURACY OF DECISION TREE CLASSIFIER : **85.21%**
* ACCURACY OF RANDOM FOREST CLASSIFIER : **85.27%**
* ACCURACY OF KNN CLASSIFIER : **78.1%**
* ACCURACY OF LOGISTIC REGRESSION : **85.8%**
* ACCURACY OF SVM CLASSIFIER : **79.7%**
---
---
**MISCLASSIFICATION PERCENTAGE:**


*   DECISION TREE CLASSIFIER : **14.78%**
*   RANDOM FOREST CLASSIFIER : **14.72%**
*   KNN CLASSIFIER : **21.88%**
*   LOGISTIC REGRESSION : **14.14%**
*   SVM CLASSIFIER: **20.25%**





---
* **ACCORDING TO ME, EXCEPT FOR KNN AND SVM CLASSIFIER, ALL MODELS ARE HAVING ACCURACY IN A CLOSE PROXIMITY.**


* ***LOGISTIC REGRESSION*** **HAS THE HIGHEST ACCURACY OF ALL. HOWEVER, BOTH DECISION TREE CLASSIFIER AND RANDOM FOREST CLASSIFIER ARE ALSO NOT TOO FAR.**

* **HOWEVER, ACCORDING TO ME, SEEING THE INTERNAL MECHANISM, RANDOM FOREST CLASSIFIER BECOMES A GREAT CHOICE**

